Sources/SwiftyUserDefaults.swift

```diff
@@ -467,33 +442,66 @@ extension UserDefaults {
 // MARK: NSCoding
 
 extension UserDefaults {
-    // TODO: Can we simplify this and ensure that T is NSCoding compliant?
     
-    public func archive<T>(_ key: DefaultsKey<T>, _ value: T) {
-        if let value: AnyObject = value as? AnyObject {
-            set(key, NSKeyedArchiver.archivedData(withRootObject: value))
-        } else {
-            assertionFailure("Invalid value type, needs to be a NSCoding-compliant type")
+    public func archive<T: NSCoding>(_ key: DefaultsKey<T>, _ value: T) {
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func archive<T: NSCoding>(_ key: DefaultsKey<T?>, _ value: T?) {
+        guard let value = value else {
+            remove(key)
+            return
         }
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[T]>, _ value: [T]) {
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
     }
     
-    public func archive<T>(_ key: DefaultsKey<T?>, _ value: T?) {
-        if let value: AnyObject = value as? AnyObject {
-            set(key, NSKeyedArchiver.archivedData(withRootObject: value))
-        } else if value == nil {
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[T]?>, _ value: [T]?) {
+        guard let value = value else {
             remove(key)
-        } else {
-            assertionFailure("Invalid value type, needs to be a NSCoding-compliant type")
+            return
         }
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[String: T]>, _ value: [String: T]) {
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
     }
     
-    public func unarchive<T>(_ key: DefaultsKey<T?>) -> T? {
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[String: T]?>, _ value: [String: T]?) {
+        guard let value = value else {
+            remove(key)
+            return
+        }
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func unarchive<T: NSCoding>(_ key: DefaultsKey<T?>) -> T? {
         return data(forKey: key._key).flatMap { NSKeyedUnarchiver.unarchiveObject(with: $0) } as? T
     }
     
-    public func unarchive<T>(_ key: DefaultsKey<T>) -> T? {
+    public func unarchive<T: NSCoding>(_ key: DefaultsKey<T>) -> T? {
         return data(forKey: key._key).flatMap { NSKeyedUnarchiver.unarchiveObject(with: $0) } as? T
     }
+    
+    public func unarchive<T: NSCoding>(_ key: DefaultsKey<[T]?>) -> [T]? {
+        return data(forKey: key._key).flatMap { NSKeyedUnarchiver.unarchiveObject(with: $0) } as? [T]
```

Comment:
I had to add this because this test wouldn't even compile: https://github.com/radex/SwiftyUserDefaults/blob/swift3/Tests/SwiftyUserDefaultsTests.swift#L538-L545 and https://github.com/radex/SwiftyUserDefaults/blob/swift3/Tests/TestHelper.swift#L23

I'm not exactly sure how it even compiled before, maybe `[NSColor]` was implicitly converted to `NSDictionary<NSColor>` and was handled here: https://github.com/radex/SwiftyUserDefaults/blob/swift3/Sources/SwiftyUserDefaults.swift#L472